Add blog post on Ionic and JHipster (#1031)
oktadeveloper/okta-blog: _source/_posts/2022-05-15-jhipster-ionic-spring-boot.adoc

---
layout: blog_post
title: "Integrate Ionic and Spring Boot Securely with JWT and OAuth 2.0"
author: matt-raible
by: advocate
communities: [mobile,java,javascript]
description: "Build an Ionic frontend for your Spring Boot backend, all using the power of JHipster!"
tags: [ionic, spring-boot, spring-security, jhipster]
tweets:
- "Create an Ionic app with a Spring Boot backend in minutes with the new Ionic blueprint for @jhipster!"
- "Introducing JHipster Ionic: 'Wow! Did you really just build that?!'"
- "Ionic is awesome. Learn how you to create apps quickly using this powerful framework. 💪"
image:
type: conversion
github: https://github.com/oktadev/okta-jhipster-ionic-example
---
:page-liquid:
:toc: macro
:experimental:
:commandkey: &#8984;

[comment]
--
todo: research and figure out best title.
Other options:

> - Integrate Ionic and Spring Boot with OAuth/OIDC

Reviewer: score: 56 neutral

> - Integrate Ionic and Spring Boot

Reviewer: Integrate Ionic and Spring Boot with the New JHipster Blueprint - 57 neutral 
Integrate Ionic and Spring Boot - 47 negative
How to Build a Mobile App with Ionic, Spring Boot, and the New JHipster Blueprint 56 - neutral (is this an accurate description?)

> - Ionic for JHipster

Reviewer: 38 negative

--

Ionic allows you to create mobile apps with web technologies that look and act like native apps.

> Because they're built with web technologies (HTML, JavaScript, and CSS), you can also deploy your Ionic apps as webapps.

Reviewer: ```suggestion
Because they're built with web technologies (HTML, JavaScript, and CSS), you can also deploy your Ionic apps as web apps.
```
web app still looks wrong to me as one word, but YMMV...

Once you've written your app with one of these, you can deploy it to a simulator or device with https://capacitorjs.com/[Capacitor].
Capacitor (pictured on the right as the blue layer) provides the runtime for your app to communicate with the native operating system and vise versa.

Ionic's main competitors are native apps built with Swift or Objective-C (for iOS) and Java or Kotlin (for Android).
Ionic also competes with React Native, which uses web technologies and translates them to native components.

The Ionic blog has a recent post that https://ionicframework.com/blog/ionic-vs-react-native-performance-comparison/[does a performance comparison between Ionic and React Native].
TL;DR: Both options will give you a high performance app with a truly native look and feel.

**Prerequisites:**

- https://nodejs.org[Node 16]
- https://sdkman.io/[Java 11]
- https://docs.docker.com/desktop/#download-and-install[Docker Desktop]

toc::[]

== ✨ Introducing the JHipster Ionic blueprint!

The JHipster project has provided support for generating an Ionic app using a `generator-jhipster-ionic` module for the past several years.

> As the primary maintainer of this module, I'm proud to announce that this module has been re-written as a blueprint and it's _much_ easier to understand now.

Reviewer: ```suggestion
As the primary maintainer of this module, I'm proud to announce that it has been re-written as a blueprint and it's _much_ easier to understand now.
```

> The previous module relied on the Ionic CLI, the base Angular https://github.com/ionic-team/starters[starter], a https://github.com/oktadev/ionic-jhipster-starter[Ionic JHipster starter], and custom code to glue it all together.

Reviewer: ```suggestion
The previous module relied on the Ionic CLI, the base Angular https://github.com/ionic-team/starters[starter], the https://github.com/oktadev/ionic-jhipster-starter[Ionic JHipster starter], and custom code to glue it all together.
```
Or should it be "an Ionic JHipster starter" ? Your call.

Author: I think "the" is OK.

Now, the source code is all contained in one project.

https://github.com/mshima[Marcelo Shima] volunteered to do the conversion and after a couple of months, I'm proud to say the https://github.com/jhipster/generator-jhipster-ionic[JHipster Ionic blueprint] is now available!
Here's how to use it:

Create a `ionic-app` directory alongside your JHipster app.

[source,shell]
----
- backend
- ionic-app
----

Navigate into `ionic-app` using your terminal.
Install Ionic for JHipster and create a new app using `jhipster-ionic`.

[source,shell]
----
npm install -g generator-jhipster-ionic
jhipster-ionic
----

You'll be prompted for the location of your JHipster app, a name for your Ionic app, and then you'll be off to the races!

image::{% asset_path 'blog/jhipster-ionic-blueprint/jhipster-ionic.png' %}[alt=JHipster Ionic prompts,width=800,align=center]

> You can also create a JHipster app and an Ionic app at the same time.
> The command will only work if you've already installed JHipster with `npm install -g generator-jhipster`.
> 
> [source,shell]
> ----
> mkdir bug-tracker && cd bug-tracker
> jhipster jdl bug-tracker.jh --blueprints ionic

Reviewer 2: Ionic blueprint has a built in JHipster, you can use it for simplicity.

```suggestion
You can also create a JHipster app and an Ionic app at the same time by using the bundled JHipster.

[source,shell]
----
mkdir bug-tracker && cd bug-tracker
jhipster-ionic jdl bug-tracker.jh
cd ../ionic4j
```

----

This will follow the same convention where the generated backend and frontend apps are side-by-side on your hard drive.

Then you can run both apps from your Ionic app using easy-to-remember commands.

[source,shell]
----
npm run backend:start
# open a new terminal window
npm start
----

image::{% asset_path 'blog/jhipster-ionic-blueprint/ionic-serve.png' %}[alt=Ionic serve command with backend running,width=800,align=center]

NOTE: The JHipster Ionic blueprint currently only supports Angular.
Now that it's a blueprint, it will be much easier to add support for Vue and React.
If you're interested in helping out, please let me know!

> Okta is a platinum sponsor of the project and enjoys assigning bug bounties for feature development.

Reviewer: ```suggestion
Okta is a platinum sponsor of the JHipster project and enjoys assigning bug bounties for feature development.
```
avoids ambiguity

> == Build a Mobile App with Ionic

Reviewer: ```suggestion
== Build a mobile app with Ionic
```
sentence case

    <td><img src="{% asset_path 'blog/jhipster-ionic-blueprint/ionic-keycloak.png' %}" alt="Ionic auth with Keycloak" width="400" loading="lazy" /></td>
  </tr>
  <tr style="background: #fff">
    <td><img src="{% asset_path 'blog/jhipster-ionic-blueprint/ionic-home.png' %}" alt="Ionic home after log in" width="400" loading="lazy" /></td>
    <td><img src="{% asset_path 'blog/jhipster-ionic-blueprint/ionic-photo.jpg' %}" alt="Hefe the Bus!" width="400" loading="lazy" /></td>
  </tr>
</table>
++++

Please keep reading to learn how JHipster made all of this possible.
Or, skip ahead to link:#run-your-ionic-app-on-ios-using-capacitor[run your Ionic app on iOS using Capacitor].

== How to integrate Ionic and Spring Boot

JHipster makes it easy to create a Spring Boot API that's protected by Spring Security.
The JHipster Ionic blueprint generates an Ionic client that talks to your Spring Boot API and understands its auth mechanism.
I created the `jhipster-ionic` project using the following steps:

Install JHipster 7.8.1 and the JHipster Ionic blueprint:

// todo: test with generator-jhipster-ionic@8.0.0
[source,shell]
----

> npm i -g generator-jhipster@7.8.1 generator-jhipster-ionic@8.0.0

Reviewer 2: JHipster is not required
```suggestion
Install the JHipster Ionic blueprint:

// todo: test with generator-jhipster-ionic@8.0.0
[source,shell]
----
npm i -g generator-jhipster-ionic@8.0.0
```

----

Create a parent directory to hold everything:

[source,shell]
----
# take is a shortcut for mdkir && cd
take jhipster-ionic
----

Clone an existing JHipster Flickr example:

[source,shell]
----
git clone https://github.com/oktadev/auth0-full-stack-java-example.git backend --depth 1
----

Create a new directory to hold your Ionic project, then run `jhipster-ionic` in it:

[source,shell]
----
take ionic-app
jhipster-ionic
----

Provide the path to your backend JHipster app and name your app `flickr2`.

// todo: update this image when JHipster Ionic blueprint is released

image::{% asset_path 'blog/jhipster-ionic-blueprint/jhipster-ionic-flickr2.png' %}[alt=JHipster Ionic with Flickr2 app,width=800,align=center]

That's it!
The blueprint will generate an Ionic client, complete with screens for editing entities, unit tests, and end-to-end tests with Cypress.

Pretty slick, don't you think?!
😎

=== Run your Spring Boot API

You'll need to start your backend first, so your Ionic app can talk to its API.
First, start Keycloak and PostgreSQL in Docker containers:

[source,shell]
----
cd backend
docker-compose -f src/main/docker/keycloak.yml up -d

> docker-compose -f src/main/docker/postgresql.yml up -d

Reviewer 2: You may want to  synchronize with line 133

> docker-compose -f src/main/docker/keycloak.yml up -d
> docker-compose -f src/main/docker/postgresql.yml up -d

Author: ```suggestion
npm run ci:e2e:prepare # starts Keycloak and PostgreSQL in Docker
```

----

> Next, update `backend/src/main/resources/config/application-prod.yml` to allow CORS from `\http://localhost:8100`.

Reviewer 2: May be relevant, if the application is generated using `jhipster-ionic jdl` or `jhipster --blueprints ionic` we inject the config.
https://github.com/jhipster/generator-jhipster-ionic/blob/61958708cd77c70704dc1b5c850d80181fe7e644/generators/server/generator.mjs#L20-L45

This was needed to make CI pass.

Author: Thanks for the tip. That looks like it'll only work if I start the app in Docker and use the `ionic-dev` profile. I'll leave as-is for this post.

</activity>
----

Modify the JHipster app's CORS settings to allow `\http://localhost` as an origin.

[source,yaml]
----
jhipster:
  ...
  cors:
    allowed-origins: 'http://localhost:8100,capacitor://localhost,http://localhost'
----

Restart your backend app and run your Ionic app on Android using the Capacitor CLI:

[source,shell]
----
npx cap run android
----

CAUTION: If you get an error when running this command, make sure you're using Java 11.

You'll need to run a couple commands to allow the emulator to communicate with JHipster and Keycloak.

[source,shell]
----
adb reverse tcp:8080 tcp:8080
adb reverse tcp:9080 tcp:9080
----

You should be able to log in and edit entities, just like you can in a browser and on iOS!

image::{% asset_path 'blog/jhipster-ionic-blueprint/flickr2-android.png' %}[alt=Flickr2 running on Android,width=300,align=center]

== Why use OpenID Connect for mobile apps?

Storing link:/blog/2019/01/22/oauth-api-keys-arent-safe-in-mobile-apps[API keys and secrets in mobile apps is not safe].
OAuth 2.0 solves this problem by not shipping any secrets in mobile apps, and instead involving the user in the process of getting an access token into the app.

> These access tokens are unique per user and every time they log in.

Reviewer: ```suggestion
These access tokens are unique per user and they're updated every time the user logs in.
```
Without this edit, the sentence feels a little ambiguous to me. Feel free to tweak as needed.
